decryptor/postgresql/pg_decryptor.go

```diff
@@ -214,89 +234,134 @@ func (proxy *PgProxy) ProxyClientConnection(errCh chan<- error) {
+		// After tha packet has been observed and possibly modified, forward it to the database.
+		if err := packet.sendPacket(); err != nil {
+			logger.WithField(logging.FieldKeyEventCode, logging.EventCodeErrorNetworkWrite).
+				WithError(err).Errorln("Can't send packet")
+			errCh <- err
+			return
 		}
-		if changed {
-			packet.ReplaceQuery(newQuery.Query())
+		// If this is a termination packet, we're done here. Signal EOF and stop the proxy.
+		if packet.terminatePacket {
+			errCh <- io.EOF
+			return
 		}
+	}
+}
 
-		censorSpan.End()
+func (proxy *PgProxy) handleClientPacket(packet *PacketHandler, logger *log.Entry) (bool, error) {
+	// Let the protocol observer take a look at the packet, keeping note of it.
+	err := proxy.protocolState.HandleClientPacket(packet)
+	if err != nil {
+		return false, err
+	}
+	switch proxy.protocolState.LastPacketType() {
+	case SimpleQueryPacket, ParseStatementPacket:
+		// If that's some sort of a packet with a query inside it,
+		// process inline data if necessary and remember the query to handle future response.
+		return proxy.handleQueryPacket(packet, logger)
+
+	case BindStatementPacket:
+		// Bound query parameters may contain inline data that we need to process.
+		// Also, remember the requested portal name for future data queries.
+		return proxy.handleBindPacket(packet, logger)
+
+	default:
+		// Forward all other uninteresting packets to the database without processing.
+		return false, nil
+	}
+}
 
-		if err := packet.sendPacket(); err != nil {
-			logger.WithField(logging.FieldKeyEventCode, logging.EventCodeErrorNetworkWrite).WithError(err).Errorln("Can't send packet")
-			errCh <- err
-			return
+func (proxy *PgProxy) handleQueryPacket(packet *PacketHandler, logger *log.Entry) (bool, error) {
+	query := proxy.protocolState.PendingQuery()
+
+	// Log query text -- if and only if we're in debug mode -- without inserted value data.
+	// The query can still be sensitive though, so only in debug mode can we do this.
+	if logging.GetLogLevel() == logging.LogDebug {
+		_, queryWithHiddenValues, _, err := common.HandleRawSQLQuery(query.Query())
+		if err == common.ErrQuerySyntaxError {
+			logger.WithError(err).WithField(logging.FieldKeyEventCode, logging.EventCodeErrorCensorQueryParseError).
+				Debugf("Parsing error on query: %s", queryWithHiddenValues)
+		} else {
+			log := logger.WithField("sql", queryWithHiddenValues)
+			if proxy.protocolState.LastPacketType() == ParseStatementPacket {
+				preparedStatement := proxy.protocolState.PendingParse()
+				log = log.WithField("prepared_name", preparedStatement.Name())
+			}
+			log.Debugln("New query")
 		}
 	}
+
+	// Let AcraCensor take a look at the query text.
+	// If it's not okay (and we're still alive), don't let the database see the query.
+	if censorErr := proxy.censor.HandleQuery(query.Query()); censorErr != nil {
```

Comment:
what about to make censor as handler for `OnQuery` subscriber for unifying? It will allow us to list all handlers which we should turn on/off for acra. And provide way to change order of handlers, their amount, etc. Or even add them as plugins by config file. It will be easier if they will be common.

Reply:
> what about to make censor as handler for OnQuery subscriber for unifying?

Hm... Sounds reasonable, actually, but we'd need to find a way to signal this special condition of AcraCensor emitting an error, but handling this error in a special way: aborting the client request and continuing service, as opposed to closing the session on other errors. I'm not sure what would be the best approach here. Downcasting to AcraCensor interface type to check for AcraCensor error specifically?.. Anyway, I'll keep a note of this idea.

> It will allow us to list all handlers which we should turn on/off for acra.

But isn't that already done via the configuration file, albeit implicitly? Or do you mean some global blade switch to enable/disable, say, AcraCensor, or tokenization, or anything else, *regardless* of the encryptor configuration file?

Reply:
> But isn't that already done via the configuration file, albeit implicitly? Or do you mean some global blade switch to enable/disable, say, AcraCensor, or tokenization, or anything else, regardless of the encryptor configuration file?

Our encryption/decryption logic works as chain of handlers which pre-defined at start and their order have some sense. For example we can detokenize at first and then decrypt. Or decrypt first and then detokenize. Or we can mask and tokenize and vice versa. And we can setup it from configuration file and allow users to configure it. And we can add censor here too. And turn-off some of these handlers. It higher level than encryption config. Because encryptor's config tells what and how to encrypt. But we can provide the order of global handlers at start. Remove them, repeat, etc.

Reply:
> For example we can detokenize at first and then decrypt. Or decrypt first and then detokenize. Or we can mask and tokenize and vice versa. And we can setup it from configuration file and allow users to configure it. And we can add censor here too.

Oh, right :+1: *That* is not possible with the current configuration file, it's all hardwired in the proxy setup code and the implementations of query observers.